### PR TITLE
feat(rs): associate & commute in expression simplification

### PR DESCRIPTION
**Pull request by kalzoo**
Adds rules to the `egg` ruleset with tests to validate behavior.

Closes https://github.com/rigetti/quil-rs/issues/244

### CODE REVIEW

File: quil-rs/src/expression/simplification.rs [hunk 1]
```diff
@@ -476,6 +476,8 @@ static RULES: Lazy<Vec<Rewrite>> = Lazy::new(|| {
         // + & *
         rw!("distribute"    ; "(* ?a (+ ?b ?c))"        => "(+ (* ?a ?b) (* ?a ?c))"),
         rw!("factor"        ; "(+ (* ?a ?b) (* ?a ?c))" => "(* ?a (+ ?b ?c))"),
+        rw!("commute-add"   ; "(+ (+ ?a ?b) ?c)"        => "(+ ?a (+ ?b ?c))"),
```
Comment (BatmanAoD) [inline quil-rs/src/expression/simplification.rs:479]: Nit: this is associativity, not commutativity. Commutativity would be `(+ ?a ?b)` => `(+ ?b ?a)`.

**Review by BatmanAoD — APPROVED** [review]

File: quil-rs/src/expression/mod.rs [hunk 2]
```diff
@@ -1046,6 +1046,13 @@ mod tests {
             ("pi", Expression::Number(PI.into())),
             ("pi/2", Expression::Number((PI / 2.0).into())),
             ("pi * pi", Expression::Number((PI.powi(2)).into())),
+            (
+                "(a[0]*2*pi)/6.283185307179586",
```
Comment (genos) [inline quil-rs/src/expression/mod.rs:1050]: 👍

File: quil-rs/src/expression/simplification.rs (hunk 1 above)
Comment (genos) [reply quil-rs/src/expression/simplification.rs:479]: ⬆️

Comment (genos) [reply quil-rs/src/expression/simplification.rs:479]: Same below

File: quil-rs/src/expression/simplification.rs [hunk 3]
```diff
@@ -563,4 +565,22 @@ mod tests {
         &RULES,
         "(^ (neg 9.48e42i) A[9])" => "(^ -9.48e42i A[9]))"
     }
+
+    egg::test_fn! {
```
Comment (genos) [inline quil-rs/src/expression/simplification.rs:569]: Great tests!

**Review by genos — COMMENTED** [review]
Agree with @BatmanAoD; `(◊ (◊ ?a ?b) ?c) == (◊ ?a (◊ ?b ?c))` (where `◊` here is one of `{+, *}`) is associativity, not commutativity. Other than the rule naming, LGTM!

If you do put in commutativity with `(◊ ?a ?b) => (◊ ?b ?a)`, you could conceivably drop a number of other rules; for instance, we'd arguably only need one of `"add-zero"` and `"zero-add"`, as we'd get the other one "for free."

File: quil-rs/src/expression/mod.rs (hunk 2 above)
Comment (Shadow53) [reply quil-rs/src/expression/mod.rs:1050]: Does this also cover `2*pi*a[0]`? Also curious if this can be generalized into `a[0]*x/x`

Comment (genos) [reply quil-rs/src/expression/mod.rs:1050]: I think there may be a way to put that together with the current rules + the changes here, but maybe adding commutativity (& trimming the ruleset, see my previous comment) might make sense.

As for your `a[0]*x/x` example, I think the `"div-mul-2"` rule covers it as long as `x` is nonzero:

```
rw!("div-mul-2" ; "(/ (* ?b ?a) ?a)" => "?b" if is_not_zero("?a")),
```

File: quil-rs/src/expression/simplification.rs (hunk 1 above)
Comment (kalzoo) [reply quil-rs/src/expression/simplification.rs:479]: doh. fixed

File: quil-rs/src/expression/simplification.rs (hunk 3 above)
Comment (kalzoo) [reply quil-rs/src/expression/simplification.rs:569]: thanks!

**Review by BatmanAoD — APPROVED** [review]

**Comment by BatmanAoD** [comment]
We're gettin' all the axioms in here

**Pull request merged**